ee/clickhouse/queries/related_actors_query.py

```diff
@@ -0,0 +1,125 @@
+import json
+from datetime import timedelta
+from functools import cached_property
+from typing import Dict, List, Literal, Optional, TypedDict, Union
+
+from django.utils.timezone import now
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.person import GET_TEAM_PERSON_DISTINCT_IDS
+from posthog.models.filters.utils import validate_group_type_index
+from posthog.models.group_type_mapping import GroupTypeMapping
+from posthog.models.property import GroupTypeIndex
+
+DISTINCT_IDS_JOIN = f"JOIN ({GET_TEAM_PERSON_DISTINCT_IDS}) pdi on e.distinct_id = pdi.distinct_id"
+
+
+class RelatedActorsResponse(TypedDict):
+    type: Literal["person", "group"]
+    group_type_index: Optional[GroupTypeIndex]
+    id: str
+    person: Optional[Dict]
+
+
+class RelatedActorsQuery:
+    """
+    This query calculates other groups and persons that are related to a person or a group.
+
+    Two actors are considered related if they have had shared events in the past 90 days.
+    """
+
+    def __init__(self, team_id: int, group_type_index: Optional[Union[GroupTypeIndex, str]], id: str):
+        self.team_id = team_id
+        self.group_type_index = validate_group_type_index("group_type_index", group_type_index)
+        self.id = id
+
+    def run(self) -> List[RelatedActorsResponse]:
+        results = self._query_related_people()
+        for group_type_mapping in GroupTypeMapping.objects.filter(team_id=self.team_id):
```

Comment:
Q: Do you want to run this even when `RelatedActorsQuery` is not passed a `group_type_index`?

i.e. when on a Persons page, would you expect to see groups here?

Reply:
Not sure I understood the question.

Yes, we show a "related groups" tab under persons with groups that share events, so this particular query is valid.